notebook/services/contents/fileio.py

```diff
@@ -249,11 +259,24 @@ def _read_notebook(self, os_path, as_version=4):
             try:
                 return nbformat.read(f, as_version=as_version)
             except Exception as e:
+                e_orig = e
+
+            # If use_atomic_writing is enabled, we'll guess that it was also
+            # enabled when this notebook was written and look for a valid
+            # atomic intermediate.
+            tmp_path = path_to_intermediate(os_path)
+
+            if not self.use_atomic_writing or not os.path.exists(tmp_path):
                 raise HTTPError(
                     400,
-                    u"Unreadable Notebook: %s %r" % (os_path, e),
+                    u"Unreadable Notebook: %s %r" % (os_path, e_orig),
                 )
 
+            # Move the bad file aside, restore the intermediate, and try again.
+            copy2_safe(os_path, path_to_invalid(os_path))
```

Comment:
I think this should be a rename rather than a copy. Copy has to open both files, read the data and write it back again, and it can fail if the disk is full. Rename changes the name of the file in the directory without reading it, and I don't think it can fail because of a full disk.

However, `os.rename()` fails on Windows if the destination file already exists. On Python 3, there's [os.replace](https://docs.python.org/3/library/os.html#os.replace) which does what we need, but we'll need to remove the destination file ourselves on Windows + Python 2.